### PR TITLE
[Gui] Issue #3923 - Add arbitrary Tree View item ordering

### PR DESCRIPTION
**Pull request by pavltom**
This pull request gives the user a possibility to reorder the document tree view items.

For more implementation details please see my contribution to the dedicated forum topic:

[Ticket #3923 - change order in tree view](https://forum.freecadweb.org/viewtopic.php?f=3&t=35316)

### CODE REVIEW

**Comment by adrianinsaval** [comment]
Compiles and runs but there are issues with the implementation I have commented here: https://forum.freecadweb.org/viewtopic.php?f=3&t=35316&p=488236#p488236

**Comment by ickby** [comment]
The viewprovider tree rank is neitehr readable, writable or observable by document observer from python. As it is a defining attribute for the viewprovider this needs to change. 
Maybe implementing it as hidden property makes it easier and more consistent within freecad?

**Comment by thebravoman** [comment]
For what it is worth, I've been in the CAD/Engineering world for decade, I installed FreeCAD and the first thing I did is add two parts and tried to reorder the parts in the tree viewer.

Looking forward to this PR

**Comment by ickby** [comment]
Thanks for the changes, looks fine for me! Two minor things that could be improved: 
1. Use the new functionality for App::Group and App::Part as those are most likely the two candidate objects for the new API
2. Write a test for the functionality to check if reordering works and is prevented correctly if it should not work

**Comment by pavltom** [comment]
Hi, I have used the possibility to disable children movement for basic document objects like Body (and all OriginGroupExtensions), Origin document object, Part Boolean object, Loft, Sweep (Pipe) and Thickness features both in Part and PartDesign. For the App::DocumentObjectGroup object, movement remains allowed, I do not see any reason why it should be restricted. Nevertheless, I am pretty much sure there are lots of other places, where moving children is not desirable, however I would let apply this to their respective maintainers. 
As for the tests, I have added a TreeView.py test, which tests the functionality both for items allowing and restricting child item movement. I believe this test is sufficient, but for sure it can be expanded to other item types or different scenarios.

**Comment by berndhahnebach** [comment]
Following a link to the branch on the CI-repository:

https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_5060

The CI-status is available on the latest commit of the branch.
If there is no status available the PR should be rebased on latest master.
Check pipeline branches to see if your PR has been run by the CI.

https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches

**Comment by freecadci** [comment]
<a href="https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_4626">pipeline status</a> for feature branch [PR_4626](https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_4626). Pipeline [#380645184 ](https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines/380645184) was triggered at [214972f](https://github.com/FreeCAD/FreeCAD/pull/4626/commits/214972ffb5529a9747a32c7f1f61ea9b4a44866b). All CI branch [pipelines](https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches).

File: src/Gui/CommandView.cpp [hunk 1]
```diff
@@ -3684,6 +3684,124 @@ void StdTreeDrag::activated(int)
...
+//===========================================================================
+// Std_GroupMoveUp
+//===========================================================================
+DEF_STD_CMD_A(StdGroupMoveUp)
+
+StdGroupMoveUp::StdGroupMoveUp()
+  : Command("Std_GroupMoveUp")
+{
+    sGroup       = QT_TR_NOOP("TreeView");
+    sMenuText    = QT_TR_NOOP("Move Up in Group");
```
Comment (chennes) [inline src/Gui/CommandView.cpp:3696]:
```suggestion
    sMenuText    = QT_TR_NOOP("Move up in group");
```

File: src/Gui/CommandView.cpp [hunk 2]
```diff
@@ -3684,6 +3684,124 @@ void StdTreeDrag::activated(int)
...
+//===========================================================================
+// Std_GroupMoveDown
+//===========================================================================
+DEF_STD_CMD_A(StdGroupMoveDown)
+
+StdGroupMoveDown::StdGroupMoveDown()
+  : Command("Std_GroupMoveDown")
+{
+    sGroup       = QT_TR_NOOP("TreeView");
+    sMenuText    = QT_TR_NOOP("Move Down in Group");
```
Comment (chennes) [inline src/Gui/CommandView.cpp:3755]:
```suggestion
    sMenuText    = QT_TR_NOOP("Move down in group");
```

**Review by chennes — CHANGES_REQUESTED** [review]
Nice addition! Thanks especially for adding the test cases. I left a couple tiny comments on the code (which overall looks really solid), and some feedback on the overall behavior.

When I move an expanded group _up_ the list, it stays open, but when I move it _down_, it collapses. This is also true if I move another object "across" the open group: it collapses if the motion is upward, but stays open if the motion is down. I personally prefer the stay-open behavior, but at the least I think it should be the same in all cases.

I also think that the multiple-selection behavior is counter-intuitive. When I have multiple objects selected and ask them to move, I felt like I was asking ALL of them to move, and then only one of them did.

Otherwise it looks to me like this is pretty much ready to merge. Thanks!

**Comment by chennes** [comment]
Thanks for the changes -- I will re-review shortly.

**Comment by chennes** [comment]
@pavltom It appears this is causing some problems, can you please have a look at https://forum.freecadweb.org/viewtopic.php?f=3&t=63744 ? Thanks.

**Pull request merged**